// node-thumbnail/src/next.js
import fs from 'fs';
import path from 'path';
import crypto from 'crypto';
import os from 'os';

import im from 'imagemagick';
import async from 'async';
import _ from 'underscore';

var options, queue, done;

const defaults = {
  prefix : '',
  suffix: '_thumb',
  digest: false,
  hashingType: 'sha1',
  width: 800,
  concurrency: os.cpus().length,
  quiet: false,
  overwrite: false,
  logger: (message) => {
    console.log(message); // eslint-disable-line no-console
  }
};

const extensions = [
  '.jpg',
  '.jpeg',
  '.JPG',
  '.JPEG',
  '.png',
  '.PNG',
  '.gif',
  '.GIF'
];


const createQueue = (settings, resolve, reject) => {

  queue = async.queue((task, callback) => {

    if (settings.digest) {

      const hash = crypto.createHash(settings.hashingType);
      const stream = fs.ReadStream(task.options.srcPath);

      stream.on('data', (d) => hash.update(d));

      stream.on('end', () => {
        const d = hash.digest('hex');

        task.options.dstPath = settings.destination + '/' + d + '_' +
          settings.width + path.extname(task.options.srcPath);

        if (settings.overwrite || !fs.existsSync(task.options.dstPath)) {
          im.resize(task.options, callback);
        }

      });

    } else {
      const name = task.options.srcPath;
      const ext = path.extname(name);
      const base = task.options.basename || path.basename(name, ext);

      task.options.dstPath = settings.destination + '/' + settings.prefix + base +
        settings.suffix + ext;

      if (settings.overwrite || !fs.existsSync(task.options.dstPath)) {
        im.resize(task.options, callback);
      }
    }

  }, settings.concurrency);

  queue.drain = () => {
    if (done) {
      done();
    }

    resolve();

    if (!settings.quiet) {
      settings.logger('All items have been processed.');
    }
  };
};


const run = (settings, resolve, reject) => {
  var images;

  if (fs.statSync(settings.source).isFile()) {
    images = [path.basename(settings.source)];
    settings.source = path.dirname(settings.source);
  } else {
    images = fs.readdirSync(settings.source);
  }

  images = _.reject(images, function(file) {
    return _.indexOf(extensions, path.extname(file)) === -1;
  });

  createQueue(settings, resolve, reject);

  _.each(images, function(image) {

    options = {
      srcPath: settings.source + '/' + image,
      width: settings.width,
      basename: settings.basename
    };

    queue.push({options: options}, function() {
      if (!settings.quiet) {
        settings.logger(image);
      }
    });

  });
};


export let thumb = (options, callback) => {
  return new Promise((resolve, reject) => {
    const settings = _.defaults(options, defaults);

    // options.args is present if run through the command line
    if (options.args) {

      if (options.args.length !== 2) {
        options.logger('Please provide a source and destination directories.');
        return;
      }

      options.source = options.args[0];
      options.destination = options.args[1];
    }

    var sourceExists = fs.existsSync(options.source);
    var destExists = fs.existsSync(options.destination);
    var errorMessage;

    if (sourceExists && !destExists) {
      errorMessage = `Destination ${options.destination} does not exist.`;
    } else if (destExists && !sourceExists) {
      errorMessage = `Source ${options.source} does not exist.`;
    } else if (!sourceExists && !destExists) {
      errorMessage = 'Source \'' + options.source + '\' and destination \'' +
        options.destination + '\' do not exist.';
      errorMessage = `Source ${options.source} and destination
        ${options.destination} do not exist.`;
    }

    if (errorMessage) {
      options.logger(errorMessage);

      if (callback) {
        callback(new Error(errorMessage));
      }

      reject(new Error(errorMessage));
    }

    if (callback) {
      done = callback;
    }

    run(settings, resolve, reject);
  });
};
